lib/reservation_manager.rb

```diff
@@ -0,0 +1,204 @@
+    def parse_reservation_data(checkin_date, checkout_date, room_number, room_block_discount = 1)
+
+      return { checkin_date: checkin_date, checkout_date: checkout_date,
+        room_number: room_number, confirmation_id: generate_random_reservation_id,
+        total_cost: calculate_total_cost(checkin_date, checkout_date, @room_cost, room_block_discount) }
+    end
+
+    #creates a reservation if there's an available room and adds to list
+    def make_reservation(checkin_date, checkout_date, num_rooms = 1, room_block_discount = 1)
+
+      num_rooms = num_rooms.to_i
+
+      if num_rooms > 2 && num_rooms < 5
+        return make_room_block(checkin_date, checkout_date, num_rooms)
+      elsif num_rooms > 5
+        raise ArgumentError.new("Room blocks of more than 5 are not allowed. Please split up reservations.") if num_rooms > 5
+      elsif num_rooms <= 0
+        raise ArgumentError.new("Please enter room numbers larger than 0.") if num_rooms
+      end
+
+      range = (Date.parse(checkin_date)..Date.parse(checkout_date)).to_a
+      room = find_available_room(range)
+
+      input = parse_reservation_data(checkin_date, checkout_date, room.room_number, room_block_discount)
+      reservation = Hotel::Reservation.new(input)
+
+      room.add_reservation(reservation)
+      add_reservation_to_list(reservation)
+      return reservation
+    end
+
+    # helper method for making room_block / parsing the room_block data
+    def parse_room_block_data(checkin_date, checkout_date, num_rooms, room_numbers)
+
+      return { checkin_date: checkin_date, checkout_date: checkout_date,
+        confirmation_id: "BLOCK#{generate_random_block_id}", num_rooms: num_rooms,
+        room_number: room_numbers, reservations: reservations, total_cost: num_rooms * (calculate_total_cost(checkin_date, checkout_date, @room_cost, @room_block_discount))
+      }
+    end
+
+    #helper method for storing a room_block/ adding it to the list
+    def add_room_block_to_list(room_block)
+      raise ArgumentError.new('Not a valid reservation.') if room_block.class != Hotel::RoomBlock
+      @room_blocks << room_block
+    end
+
+    #helper method for make reservation, which makes room block and its reservations
+    def make_room_block(checkin_date, checkout_date, num_rooms = 2)
+      reservations = []
+      room_numbers = []
+
+      num_rooms.times do
+        reservation = make_reservation(checkin_date, checkout_date, 1, @room_block_discount)
+        reservations << reservation
+        room_numbers << reservation.room_number
+      end
+
+      block_data = parse_room_block_data(checkin_date, checkout_date, num_rooms, room_numbers)
+
+      room_block =  Hotel::RoomBlock.new(block_data)
+      add_room_block_to_list(room_block)
+
+      return room_block
+    end
+
+    #returns total cost for a given reservation
+    def get_total_reservation_cost(given_confirmation_id)
+
+      return @reservations.reduce(0) do |cost, reservation|
```

Comment:
Nice use of `reduce`!

We could also probably write this with `find` (or in Rails we would use `find_by`).